pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -318,21 +375,88 @@ func (rd *restoreDataProcessor) openSSTs(
 		return nil
 	}
 
+	hashSpan := func(span roachpb.Span) (uint32, error) {
+		hash := fnv.New32a()
+		_, err := hash.Write(span.Key)
+		if err != nil {
+			return 0, err
+		}
+
+		return hash.Sum32() % uint32(rd.numWorkers), nil
+	}
+
 	log.VEventf(ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
 
-	storeFiles := make([]storageccl.StoreFile, 0, len(EntryFiles{}))
-	for _, file := range entry.Files {
+	storeFiles := make([]storageccl.StoreFile, 0, len(entry.Files))
+	iterAllocs := make([]*quotapool.IntAlloc, 0, len(entry.Files))
+	var sstOverheadBytesPerFile uint64
+	if rd.spec.Encryption != nil {
+		sstOverheadBytesPerFile = sstReaderEncryptedOverheadBytesPerFile
+	} else {
+		sstOverheadBytesPerFile = sstReaderOverheadBytesPerFile
+	}
+
+	for idx := 0; idx < len(entry.Files); {
+		file := entry.Files[idx]
 		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
 
+		alloc, err := rd.sstMemQuotaPool.TryAcquire(ctx, sstOverheadBytesPerFile)
+		if errors.Is(err, quotapool.ErrNotEnoughQuota) {
+			if err := rd.mem.Grow(ctx, int64(sstOverheadBytesPerFile)); err != nil {
+				// If we failed to allocate more memory, send the iterator
+				// containing the files we have right now.
+				if len(storeFiles) > 0 {
+					iterOpts := storage.IterOptions{
+						RangeKeyMaskingBelow: rd.spec.RestoreTime,
+						KeyTypes:             storage.IterKeyTypePointsAndRanges,
+						LowerBound:           keys.LocalMax,
+						UpperBound:           keys.MaxKey,
+					}
+					iter, err := storageccl.ExternalSSTReader(ctx, storeFiles, rd.spec.Encryption, iterOpts)
+					if err != nil {
+						return err
+					}
+					workerNum, err := hashSpan(entry.Span)
+					if err != nil {
+						return err
+					}
+
+					log.VInfof(ctx, 2, "sending iterator after %d out of %d files due to insufficient memory", idx, len(entry.Files))
+					if err := sendIter(iter, dirs, workerSSTChannels[workerNum], iterAllocs, false); err != nil {
```

Comment:
I have the workers pull off of their own channel after send iter fails to send an iterator for the entire span. I believe this is necessary since we don't want an out of order write between the first half of files and the second half of files for the same span, which is possible if all workers pull off of the same queue.

Reply:
I see why you did this, thanks, and thanks for the chat about this issue, it was helpful.

I think there are probably other ways to guarantee ordering, for example, what do you think about:
- moving the call to openSSTs into the worker itself (I was thinking of doing that anyway because we need to increase parallelism for openSSTs, either by having multiple goroutines or by moving it to the worker)
- then, each large request that needs to be fulfilled by multiple partial ssts can be done sequentially by a single worker, meaning, the worker will process the first partial sst, flush it, and then start processing the next one.